Native STOMP

ikavgo opened the pull request:
**NOTE**: Bazel tests pass locally, I don't like the code yet though. 
**TODO**: Double check Management UI correctly shows Connection information.

## Proposed Changes

Don't use AMQP 0-9-1 as backend for STOMP.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [x] Cosmetic change (whitespace, formatting, etc)
- [ ] Build system and/or CI

## Checklist

_Put an `x` in the boxes that apply.
You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask on the mailing list.
We're here to help!
This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] All tests pass locally with my changes
- [x] If relevant, I have added necessary documentation to https://github.com/rabbitmq/rabbitmq-website
- [ ] If relevant, I have added this change to the first version(s) in release-notes that I expect to introduce it

## Further Comments

Documentation update: https://github.com/rabbitmq/rabbitmq-website/pull/1713

ikvmw commented:
Regarding Connection closure, UTF, queues utilization - the idea is to merge this first, then tweak standard support and resources.

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -684,117 +690,383 @@ do_subscribe(Destination, DestHdr, Frame,
+
+            %% {ok, BasicMessage} = rabbit_basic:message(ExchangeName, RoutingKey, Content),
+
+            %% Delivery = #delivery{
```
Delete any commented code.

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -7,85 +7,101 @@
+%%   when SendFun :: fun((atom(), binary()) -> term()),
+%%        AdapterInfo :: #amqp_adapter_info{},
+%%        SSLLoginName :: atom() | binary(),
+%%        PeerAddr :: inet:ip_address().
```
Uncomment the type specs and make sure dialyzer passes.

ansd commented:
This comment is work in progress. I keep updating this comment while reviewing.

1. Do we still need the runtime dependency on AMQP 0.9.1 client in the STOMP plugin? If not, please remove that dependency from the Makefile and Bazel file. That's what was done for Native MQTT as well.

ansd commented on deps/rabbitmq_stomp/BUILD.bazel:
```diff
@@ -4,6 +4,7 @@ load("@rules_erlang//:dialyze.bzl", "dialyze", "plt")
 load(
     "//:rabbitmq.bzl",
     "BROKER_VERSION_REQUIREMENTS_ANY",
+    "ENABLE_FEATURE_MAYBE_EXPR",
```
We will remove enabling the maybe feature at runtime for 3.13 as 3.13 requires OTP 26.
However, it's needed right now.

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp.erl:
```diff
@@ -28,6 +28,7 @@
 start(normal, []) ->
     Config = parse_configuration(),
     Listeners = parse_listener_configuration(),
+    rabbit_global_counters:init([{protocol, stomp}]),
```
I'd vote to differentiate by STOMP versions, i.e. 1.0, 1.1, 1.2 since that's what we currently do for MQTT (3.1, 3.1.1, 5.0).

ansd replied:
Where are the global counters per protocol and queue type initialised? Are they missing?

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -140,21 +164,32 @@ initial_state(Configuration,
        %%       to override this value?
```
Delete the adapter info. It's not needed anymore.

ansd replied:
The management UI shows that there is 1 AMQP 0.9.1 channel for the STOMP connection, which is wrong.

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -684,117 +690,383 @@ do_subscribe(Destination, DestHdr, Frame,
+                      end, State).
+
+coalesce_and_send(MsgSeqNos, NegativeMsgSeqNos, MkMsgFun, State = #proc_state{unconfirmed = UC}) ->
+    SMsgSeqNos = lists:usort(MsgSeqNos),
+    UnconfirmedCutoff = case rabbit_confirms:is_empty(UC) of
+                 true  -> lists:last(SMsgSeqNos) + 1;
+                 false -> rabbit_confirms:smallest(UC)
+             end,
+    Cutoff = lists:min([UnconfirmedCutoff | NegativeMsgSeqNos]),
+    {Ms, Ss} = lists:splitwith(fun(X) -> X < Cutoff end, SMsgSeqNos),
+    State1 = case Ms of
+                 [] -> State;
+                 _  -> MkMsgFun(lists:last(Ms), true, State)
+             end,
+    lists:foldl(fun(SeqNo, StateN) ->
+                        MkMsgFun(SeqNo, false, StateN)
+                end, State1, Ss).
```
Where is this code tested? If it's not tested, please add tests.

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -1217,3 +1513,504 @@ maybe_apply_default_topic_exchange(Exchange, _DefaultTopicExchange) ->
+            State1 = State#proc_state{queue_states = QState1},
+            State2 = handle_queue_actions(Actions, State1),
+            {ok, State2};
+        {eol, Actions} ->
+            State1 = handle_queue_actions(Actions, State),
+            State2 = handle_consuming_queue_down_or_eol(QRef, State1),
+            {ConfirmMXs, UC1} =
+                rabbit_confirms:remove_queue(QRef, State1#proc_state.unconfirmed),
+            %% Deleted queue is a special case.
+            %% Do not nack the "rejected" messages.
+            State3 = record_confirms(ConfirmMXs,
+                                     State2#proc_state{unconfirmed = UC1}),
+            {ok, State3#proc_state{queue_states = rabbit_queue_type:remove(QRef, QStates0)}};
```
Where is this code tested? If it's not tested, please add tests.

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -1217,3 +1513,504 @@ maybe_apply_default_topic_exchange(Exchange, _DefaultTopicExchange) ->
+          ({settled, QRef, MsgSeqNos}, S0) ->
+              S = confirm(MsgSeqNos, QRef, S0),
+              send_confirms_and_nacks(S);
+          ({rejected, _QRef, MsgSeqNos}, S0) ->
+              {U, Rej} =
+              lists:foldr(
+                fun(SeqNo, {U1, Acc}) ->
+                        case rabbit_confirms:reject(SeqNo, U1) of
+                            {ok, MX, U2} ->
+                                {U2, [MX | Acc]};
+                            {error, not_found} ->
+                                {U1, Acc}
+                        end
+                end, {S0#proc_state.unconfirmed, []}, MsgSeqNos),
+              S = S0#proc_state{unconfirmed = U},
+              %% Don't send anything, no nacks in STOMP
+              record_rejects(Rej, S);
+          ({queue_down, QRef}, S0) ->
+              handle_consuming_queue_down_or_eol(QRef, S0);
+          %% TODO: I have no idea about the scope of credit_flow
+          ({block, QName}, S0) ->
+              credit_flow:block(QName),
+              S0;
+          ({unblock, QName}, S0) ->
+              credit_flow:unblock(QName),
+              S0;
+          %% TODO: in rabbit_channel there code for handling
+          %% send_drained and send_credit_reply
+          %% I'm doing catch all here to not crash?
+          (_, S0) ->
+              S0
```
Where is this code tested? If it's not tested, please add tests.

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -1217,3 +1513,504 @@ maybe_apply_default_topic_exchange(Exchange, _DefaultTopicExchange) ->
+
+
+
+parse_endpoint(undefined) ->
+    parse_endpoint("/queue");
+parse_endpoint(Destination) when is_binary(Destination) ->
+    parse_endpoint(unicode:characters_to_list(Destination));
```
Where is this code tested? If it's not tested, please add tests.

ikvmw replied:
parse_endpoint called on each SEND and SUBSCRIBE. Should be many dozens hits across our test suites

ansd replied:
I observe 0 hits for these 2 clauses ☹️

<img width="768" alt="Screenshot 2023-09-06 at 08 52 17" src="https://github.com/rabbitmq/rabbitmq-server/assets/12648310/a1e63115-518c-4bdd-a3d6-b0e918773e2e">

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -1217,3 +1513,504 @@ maybe_apply_default_topic_exchange(Exchange, _DefaultTopicExchange) ->
+    end.
+
+erase_queue_stats(QName) ->
+    rabbit_core_metrics:channel_queue_down({self(), QName}),
+    erase({queue_stats, QName}),
+    [begin
+	 rabbit_core_metrics:channel_queue_exchange_down({self(), QX}),
+	 erase({queue_exchange_stats, QX})
+     end || {{queue_exchange_stats, QX = {QName0, _}}, _} <- get(),
+	    QName0 =:= QName].
```
Why are these stats erased? It looks like there aren't written in the first place?

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -1217,3 +1513,504 @@ maybe_apply_default_topic_exchange(Exchange, _DefaultTopicExchange) ->
+
+    maybe
+        {ok, User} ?= rabbit_access_control:check_user_login(Username, AuthProps),
+        {ok, AuthzCtx} ?= check_vhost_access(VHost, User, Addr),
```
Should we additionally check for whether the vhost exists and whether the vhost is alive? (see Native MQTT).

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -1217,3 +1513,504 @@ maybe_apply_default_topic_exchange(Exchange, _DefaultTopicExchange) ->
+                            end,
+
+    maybe
+        {ok, User} ?= rabbit_access_control:check_user_login(Username, AuthProps),
```
`rabbit_core_metrics:auth_attempt_failed()` is missing if user login is refused.

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -7,85 +7,101 @@
+                     user, queue_states, delivery_tag = 0, msg_seq_no, delivery_flow,
+                     default_topic_exchange, default_nack_requeue}).
+
+-record(subscription, {dest_hdr, ack_mode, multi_ack, description, queue_name}).
+
+-record(pending_ack, {
+                      %% delivery identifier used by clients
+                      %% to acknowledge and reject deliveries
+                      delivery_tag,
+                      %% consumer tag
+                      tag,
+                      delivered_at,
+                      %% queue name
+                      queue,
+                      %% message ID used by queue and message store implementations
+                      msg_id
+                    }).
```
add type specs

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -803,65 +1075,104 @@ negotiate_version(Frame) ->
+    ExchangeNameBin = mc:get_annotation(exchange, MsgCont0),
+    MsgCont = mc:convert(mc_amqpl, MsgCont0),
+    Content = mc:protocol_state(MsgCont),
+    Delivery = #'basic.deliver'{consumer_tag = ConsumerTag,
+                                delivery_tag = DeliveryTag,
+                                redelivered  = Redelivered,
+                                exchange     = ExchangeNameBin,
+                                routing_key  = RoutingKey},
```
Creating an AMQP 0.9.1 `#basic.deliver{}` is unnecessary. The STOMP protocol is used to send the message to the client, not the AMQP 0.9.1 protocol. I think you should restructure the code, specifically `rabbit_stomp_util:headers_extra/4` should not accept a `#basic.deliver{}` anymore.

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -684,117 +690,383 @@ do_subscribe(Destination, DestHdr, Frame,
-            end;
+            {ExchangeNameList, RoutingKeyList} = parse_routing(Destination, DfltTopicEx),
+            %% io:format("Parse_routing: ~p~n", [{ExchangeNameList, RoutingKeyList}]),
+            RoutingKey = list_to_binary(RoutingKeyList),
```
The STOMP plugin operates too much on lists instead of binaries which is inefficient. At some point we should improve this. However, this should be a follow up PR.

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -7,85 +7,101 @@
-
--record(subscription, {dest_hdr, ack_mode, multi_ack, description}).
+-record(proc_state, {session_id, subscriptions,
+                     version, start_heartbeat_fun, pending_receipts,
```
My editor shows that `start_heartbeat_fun` is an unused field. Where is this field used? Remove any unused fields.

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -7,85 +7,101 @@
+                      msg_id
+                    }).
 
 -define(FLUSH_TIMEOUT, 60000).
```
remove unused macro

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_reader.erl:
```diff
@@ -20,10 +20,6 @@
 -include("rabbit_stomp_frame.hrl").
 -include_lib("amqp_client/include/amqp_client.hrl").
```
remove

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -1217,3 +1513,504 @@ maybe_apply_default_topic_exchange(Exchange, _DefaultTopicExchange) ->
+            {error, queue_limit_exceeded}
+    end.
+
+routing_init_state() -> sets:new().
```
use v2

ikvmw replied:
what is v2?

ansd replied:
https://www.erlang.org/doc/man/sets.html#description
> Erlang/OTP 24.0 introduced a new internal representation for sets which is more performant. Developers can use this new representation by passing the {version, 2} flag to [new/1](https://www.erlang.org/doc/man/sets.html#new-1) and [from_list/2](https://www.erlang.org/doc/man/sets.html#from_list-2), such as sets:new([{version, 2}]). This new representation will become the default in future Erlang/OTP versions. Functions that work on two sets, such as [union/2](https://www.erlang.org/doc/man/sets.html#union-2) and similar, will work with sets of different versions. In such cases, there is no guarantee about the version of the returned set. Explicit conversion from the old version to the new one can be done with sets:from_list(sets:to_list(Old), [{version,2}]).

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -471,64 +513,98 @@ cancel_subscription({error, _}, _Frame, State) ->
-                              #'queue.delete'{queue  = list_to_binary(QName),
-                                              nowait = false}),
+            QRes = rabbit_misc:r(VHost, queue, list_to_binary(QName)),
+            io:format("Durable QRes: ~p~n", [QRes]),
```
delete

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -1217,3 +1513,504 @@ maybe_apply_default_topic_exchange(Exchange, _DefaultTopicExchange) ->
+    case rabbit_amqqueue:with(
+           QRes,
+           fun (Q) ->
+                   io:format("Delete queue ~p~n", [rabbit_queue_type:delete(Q, false, false, Username)])
```
Use the logger if you want to log something.

ansd replied:
missing access control check for queue deletion

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -1217,3 +1513,504 @@ maybe_apply_default_topic_exchange(Exchange, _DefaultTopicExchange) ->
+    Binding = #binding{source = rabbit_misc:r(VHost, exchange, list_to_binary(Exchange)),
+                       destination = QName,
+                       key = list_to_binary(RoutingKey)},
+    case rabbit_binding:add(Binding, Username) of
```
missing access control checks

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -1217,3 +1513,504 @@ maybe_apply_default_topic_exchange(Exchange, _DefaultTopicExchange) ->
+              credit_flow:block(QName),
+              S0;
+          ({unblock, QName}, S0) ->
+              credit_flow:unblock(QName),
```
It makes probably more sense to do what has been done for Native MQTT, i.e .not using `credit_flow`.

michaelklishin replied:
Yes, not using credit flow has worked well so far for MQTT, so let's just drop it

ikvmw replied:
that's great! Flow was on my "mysteries" list

ikvmw replied:
hmm mqtt_processor still uses `credit_flow` - https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbitmq_mqtt/src/rabbit_mqtt_processor.erl#L1914.

ikavgo replied:
still used in MQTT? https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbitmq_mqtt/src/rabbit_mqtt_processor.erl#L2342

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -887,38 +1198,32 @@ ensure_reply_to(Frame = #stomp_frame{headers = Headers}, State) ->
-
-            Destination = binary_to_list(Queue),
+            Spec = #{no_ack => true,
+                     prefetch_count => application:get_env(rabbit, default_consumer_prefetch),
```
Wrong prefetch count.
It leads to crashes later on.

e.g.
```
SEND
destination:/queue/reply-test
reply-to:/temp-queue/foo

Hello World!^@
```
then click on the reply-to-queue in the Management UI. You get a 500 error and RabbitMQ logs:
```
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>   crasher:
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>     initial call: cowboy_stream_h:request_process/3
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>     pid: <0.1278.0>
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>     registered_name: []
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>     exception error: no function clause matching
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>                      thoas_encode:value({ok,{false,0}},
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>                                         #Fun<thoas_encode.0.30747453>) (src/thoas_encode.erl, line 1720)
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>       in function  thoas_encode:map_naive_loop/2 (src/thoas_encode.erl, line 1712)
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>       in call from thoas_encode:map_naive_loop/2 (src/thoas_encode.erl, line 1713)
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>       in call from thoas_encode:map_naive/2 (src/thoas_encode.erl, line 1704)
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>       in call from thoas_encode:list/2 (src/thoas_encode.erl, line 1692)
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>       in call from thoas_encode:map_naive/2 (src/thoas_encode.erl, line 1703)
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>       in call from thoas:encode/2 (src/thoas.erl, line 92)
2023-09-06 14:10:00.942131+00:00 [error] <0.1278.0>       in call from rabbit_mgmt_util:reply0/3 (rabbit_mgmt_util.erl, line 252)
```

ansd commented on deps/rabbitmq_web_stomp/src/rabbit_web_stomp_handler.erl:
```diff
@@ -233,21 +204,37 @@ websocket_info({start_heartbeats, {SendTimeout, ReceiveTimeout}},
+websocket_info(connection_created, State) ->
+    Infos = infos(?INFO_ITEMS ++ ?OTHER_METRICS, State),
+
+    ?LOG_INFO("Connection created infos ~p", [Infos]),
```
Logging the connection infos at `info` level of every Web STOMP connection pollutes the logs too much.

ansd commented on deps/rabbitmq_web_stomp/src/rabbit_web_stomp_handler.erl:
```diff
@@ -233,21 +204,37 @@ websocket_info({start_heartbeats, {SendTimeout, ReceiveTimeout}},
                                        SendFun, ReceiveTimeout, ReceiveFun),
-    {ok, State#state{heartbeat = Heartbeat}};
+    {ok, State#state{heartbeat = Heartbeat,
+                     timeout_sec = {0, 0}}};
```
`{0, 0}` is wrong?

ansd commented on deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl:
```diff
@@ -684,117 +690,383 @@ do_subscribe(Destination, DestHdr, Frame,
+                          payload_fragments_rev = [BodyFragments]
+                         },
 
+            Message0 = mc_amqpl:message(ExchangeName, RoutingKey, Content0),
```
I think it would be great to directly store STOMP messages as part of this PR and to implement a `mc_stomp` module that does the translations. It's unnecessary to convert from / to AMQP 0.9.1 message format. But it can also be done as a separate PR.

mkuratczyk commented:
Using [a simple test app](https://github.com/mkuratczyk/stomp-test), it seems like publishing is significantly slower than on `main`:

```
> time ./stomp-go -count 100000 -publishOnly -queue foo
sender finished

________________________________________________________
Executed in    2.11 secs      fish           external
```


This branch:
```
> time ./stomp-go -count 100000 -publishOnly -queue foo
sender finished

________________________________________________________
Executed in   22.91 secs      fish           external
```
(note there's a 1 second sleep in the app, to avoid premature app termination, before all messages were sent)

mkuratczyk commented:
Messages published over STOMP, can't be consumed over AMQP 0.9.1.
Using [a simple test app](https://github.com/mkuratczyk/stomp-test) for publishing and then perf-test for conuming:

```
# publish 1 message
./stomp-go -count 1 -publishOnly -queue stomp-to-amqp
# attempt to consume the messsage
perf-test -x 0 -ad false -f persistent -u stomp-to-amqp
```

logs:
```
   crasher:
     initial call: rabbit_writer:enter_mainloop/2
     pid: <0.1334.0>
     registered_name: []
     exception error: bad argument
       in function  size/1
          called as size([<<"Message #1">>])
          *** argument 1: not tuple or binary
       in call from rabbit_binary_generator:build_content_frames/7 (rabbit_binary_generator.erl, line 89)
       in call from rabbit_binary_generator:build_simple_content_frames/4 (rabbit_binary_generator.erl, line 61)
       in call from rabbit_writer:assemble_frames/5 (rabbit_writer.erl, line 334)
       in call from rabbit_writer:internal_send_command_async/3 (rabbit_writer.erl, line 365)
       in call from rabbit_writer:handle_message/2 (rabbit_writer.erl, line 265)
       in call from rabbit_writer:handle_message/3 (rabbit_writer.erl, line 232)
       in call from rabbit_writer:mainloop1/2 (rabbit_writer.erl, line 223)
     ancestors: [<0.1333.0>,<0.1323.0>,<0.1318.0>,<0.1317.0>,<0.1137.0>,
                   <0.1136.0>,<0.1135.0>,<0.1133.0>,<0.1132.0>,rabbit_sup,
                   <0.243.0>]
     message_queue_len: 0
     messages: []
     links: [<0.1333.0>]
     dictionary: [{process_name,
                       {rabbit_writer,{<<"[::1]:57240 -> [::1]:5672">>,1}}}]
     trap_exit: false
     status: running
     heap_size: 610
     stack_size: 28
     reductions: 419
   neighbours:
 
     supervisor: {<0.1333.0>,rabbit_channel_sup}
     errorContext: child_terminated
     reason: {badarg,
                 [{erlang,size,
                      [[<<"Message #1">>]],
                      [{error_info,#{module => erl_erts_errors}}]},
                  {rabbit_binary_generator,build_content_frames,7,
                      [{file,"rabbit_binary_generator.erl"},{line,89}]},
                  {rabbit_binary_generator,build_simple_content_frames,4,
                      [{file,"rabbit_binary_generator.erl"},{line,61}]},
                  {rabbit_writer,assemble_frames,5,
                      [{file,"rabbit_writer.erl"},{line,334}]},
                  {rabbit_writer,internal_send_command_async,3,
                      [{file,"rabbit_writer.erl"},{line,365}]},
                  {rabbit_writer,handle_message,2,
                      [{file,"rabbit_writer.erl"},{line,265}]},
                  {rabbit_writer,handle_message,3,
                      [{file,"rabbit_writer.erl"},{line,232}]},
                  {rabbit_writer,mainloop1,2,
                      [{file,"rabbit_writer.erl"},{line,223}]}]}
     offender: [{pid,<0.1334.0>},
                {id,writer},
                {mfargs,{rabbit_writer,start_link,
                                       [#Port<0.168>,1,131072,
                                        rabbit_framing_amqp_0_9_1,<0.1319.0>,
                                        {<<"[::1]:57240 -> [::1]:5672">>,1},
                                        true]}},
                {restart_type,transient},
                {significant,true},
                {shutdown,70000},
                {child_type,worker}]
 
     supervisor: {<0.1333.0>,rabbit_channel_sup}
     errorContext: shutdown
     reason: reached_max_restart_intensity
     offender: [{pid,<0.1334.0>},
                {id,writer},
                {mfargs,{rabbit_writer,start_link,
                                       [#Port<0.168>,1,131072,
                                        rabbit_framing_amqp_0_9_1,<0.1319.0>,
                                        {<<"[::1]:57240 -> [::1]:5672">>,1},
                                        true]}},
                {restart_type,transient},
                {significant,true},
                {shutdown,70000},
                {child_type,worker}]
```

mkuratczyk commented:
There's a heartbeat related crash when publishing more than a couple messages.

```
./stomp-go -count 10000 -publishOnly
```
leads to
```
   crasher:
     initial call: rabbit_heartbeat:'-heartbeater/2-fun-0-'/0
     pid: <0.2263.0>
     registered_name: []
     exception exit: {unexpected_message,resume}
       in function  rabbit_heartbeat:heartbeater/3 (rabbit_heartbeat.erl, line 138)
     ancestors: [<0.2260.0>,<0.2259.0>,<0.910.0>,<0.909.0>,<0.908.0>,
                   <0.906.0>,<0.905.0>,rabbit_stomp_sup,<0.903.0>]
     message_queue_len: 0
     messages: []
     links: [<0.2260.0>]
     dictionary: [{process_name,{heartbeat_receiver,unknown}}]
     trap_exit: false
     status: running
     heap_size: 376
     stack_size: 28
     reductions: 138
   neighbours:
 
     supervisor: {<0.2260.0>,rabbit_connection_helper_sup}
     errorContext: child_terminated
     reason: {unexpected_message,resume}
     offender: [{pid,<0.2263.0>},
                {id,heartbeat_receiver},
                {mfargs,
                    {rabbit_heartbeat,start_heartbeat_receiver,
                        [#Port<0.461>,60,#Fun<rabbit_stomp_reader.2.4391843>,
                         {heartbeat_receiver,unknown}]}},
                {restart_type,transient},
                {significant,false},
                {shutdown,300000},
                {child_type,worker}]
```

If I change 10000 to 10, there's no stacktrace.

NOTE: the same happens on `main`, so feel free to treat it as a separate issue.

mkuratczyk commented:
Some performance observations at this point.

100 publishers -> 100 queues -> 100 consumers (note: there's no flow control here):
` ./stomp-go -count 100000 -queue stomp-to-stomp -publisherCount 100 -consumerCount 100 -separateQueues`
main: 75s
PR: 47s

Memory usage with 1000 consumers connected (` ./stomp-go -consumeOnly -consumerCount 1000`):
main: 600MB (~17500 erlang processes)
PR: 390MB (~5500 erlang processes)

However, single connection/queue consumption performance seems degraded.
Publish with `./stomp-go -count 1000000 -publishOnly -queue one -timestampBody` takes 10-11s for both `main` and this PR. However, consuming this backlog with `./stomp-go -count 1000000 -consumeOnly -queue one -timestampBody` takes 23-24s on `main` and 30s with this PR.

ikavgo committed:
Native STOMP: remove amqp_client dependency, rewrite frame parser, binary headers

Remove the AMQP 0-9-1 client dependency from the STOMP plugin, making it a
truly native protocol implementation (following the MQTT native pattern).

Key changes:

- Remove amqp_client from DEPS; replace #amqp_adapter_info{} with #conn_info{}
  record storing connection endpoints directly via rabbit_net:socket_ends/2
- Replace #'basic.deliver'{} construction with direct parameter passing
- Add per-version global counters (STOMP 1.0/1.1/1.2) and per-queue-type
  initialization, publisher/consumer lifecycle counters
- Add force_event_refresh handler, rabbit_trace integration (tap_in/tap_out),
  consumer timeout handling, vhost existence and connection limit checks
- Add structured logger metadata (connection, vhost, user) for traceable logs
- Rewrite frame parser: bulk binary scanning for commands and headers (zero-copy
  fast path), byte-by-byte fallback only for escape sequences, bounded
  accumulation in all phases, unknown commands produce frames (not parse errors)
- Convert all header representation from charlists to binaries throughout:
  frame parser, serializer, header accessors, rabbit_stomp_util,
  rabbit_stomp_processor, and all test files
- Restructure python_SUITE: auto-generate per-function CT test cases from
  Python test files, organized into nested groups by class
- Fix rabbitmq_cli plugin tests to reflect STOMP's independence from amqp_client
- Update rabbit_web_stomp_handler for all API changes

michaelklishin committed:
Native STOMP: fix `command/2` return type for unauthenticated clients

`command/2` returned a 3-tuple `{ok, State, none}` when an unauthenticated
client sent a non-CONNECT frame. The reader expects a 2-tuple, `{ok, State}`.

michaelklishin committed:
Native STOMP: make the frame parser more robust

1. A negative content-length crashed the parser in `parse_known_body/4`.
2. A non-numeric content-length (e.g. "abc") crashed `integer_header/2`
    with a `badarg` exception
3. A content-length pointing past the NUL
   terminator resulted in a `badmatch`.

All three are pre-auth exceptions.

With these changes:

1. Negative values now return `{error, {invalid_content_length, N}}`
2. Non-numeric values are treated as absent
3. A missing NUL terminator returns `{error, missing_body_terminator}`

michaelklishin committed:
STOMP: more test suite improvements
